File: Tree/Leetcode-226-InvertBinaryTree/Leetcode-226-InvertBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        changeBoth(root);
        return root;
    }
    void changeBoth(TreeNode* subtree) {
        if( subtree == nullptr ) return;

        TreeNode* temp = subtree->right;
        subtree->right = subtree->left;
        subtree->left = temp;

        changeBoth(subtree->left);
        changeBoth(subtree->right);

    }
};

File: 2월3주차/leetcode-226-Invert Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution: # bfs
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        dq = collections.deque([root])
        while dq:
            for i in range(len(dq)):
                curnode = dq.popleft()
                if not curnode:
                    continue
                curnode.left, curnode.right = curnode.right, curnode.left
                dq.append(curnode.left)
                dq.append(curnode.right)
        return root
    
class Solution: # dfs
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]: # 자식이 서로 바뀐 root를 리턴하는 재귀함수
        if root:
            root.left, root.right = self.invertTree(root.right), self.invertTree(root.left)
            return root
        else:
            return None